Haiku: Add Haiku support for CoreCLR/PAL

trungnt2910 opened the pull request:
This contains the code required to build this repo's native components and get `paltests` to pass on Haiku.

Most implementation details related to this commit are documented [here](https://www.haiku-os.org/blog/trungnt2910/2023-05-14_gsoc_2023_dotnet_port_progress_1/).

Part of https://github.com/dotnet/runtime/issues/55803.

jkotas commented on src/native/minipal/errno.h:
```diff
@@ -0,0 +1,48 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
This file should not be needed.

CoreCLR PAL implementation should be able to use definition of these constants from Haiku system headers. I assume that all these constants are defined in Haiky system headers. Is that correct?

Components using the CoreCLR PAL should use definitions provided by the CoreCLR PAL header.

trungnt2910 replied:
I believe the discussion has been done here: https://github.com/dotnet/runtime/issues/55803#issuecomment-886096255, unless something major has changed since the .NET 6 - 7 times.

@jessicah's port (https://github.com/jessicah/dotnet-runtime) uses `#ifdef` for every `errno` wherever needed; this header gathers them all to one place.

jkotas replied:
I agree that there may be some changes required to deal with the errno given how the CoreCLR PAL is structured (https://github.com/dotnet/runtime/issues/55803#issuecomment-886696103).

All changes related to errno should be under src/coreclr/pal. No changes should be needed outside this directory, except of deleting unnecessary code. I have left some comments on the @jessicah's commit: https://github.com/jessicah/dotnet-runtime/commit/7bd4f1d40b4d938b81eecbc77d80e145e005118b

trungnt2910 replied:
Done. I've removed the `minpal/errno.h` header.

However, unlike what you said about the definitions in `sstring.cpp` (a file _outside_ `src/coreclr/pal`, located in `src/coreclr/utilcode` instead), the definitions are still needed:

https://github.com/dotnet/runtime/blob/b09e18e8b6092e35611c2567bd55e8909ea78d7d/src/coreclr/utilcode/sstring.cpp#L1794-L1803

For these files, I've replaced the `#include <minipal/errno.h>` with `#include <pal.h>`, though I don't really know what's the correct solution.

For now, nothing seems to have broken. Builds are still fine, `paltests` still passing.

trungnt2910 replied:
Oops, seems like `pal.h` is not accessible from `sstring.cpp`:

> D:\a\_work\1\s\src\coreclr\utilcode\sstring.cpp(14): fatal error C1083: Cannot open include file: 'pal.h': No such file or directory

janvorli replied:
@trungnt2910 the pal.h is not available on Windows. You should never include it anywhere from the runtime code explicitly.  It gets automatically included when runtime includes any of the standard headers. They are redirected to palrt.h and that one pulls in pal.h

janvorli commented on src/coreclr/pal/inc/rt/safecrt.h:
```diff
@@ -131,15 +131,7 @@ typedef int errno_t; /* standard */
-#if !defined(EILSEQ)
-#define EILSEQ          42
-#endif
+#include <pal.h>
```
This include should not be needed, the palrt.h includes pal.h and then it includes the safecrt.h, so pal.h should be already included here.

trungnt2910 replied:
Done.

janvorli commented on src/coreclr/utilcode/sstring.cpp:
```diff
@@ -11,6 +11,7 @@
 #include "ex.h"
 #include "holder.h"
 
+#include <pal.h>
```
I think that pal.h should have been included transitively from one of the previous includes, so this should not be needed.

trungnt2910 replied:
Done. Let's see if it fixes/breaks anything.

janvorli commented on src/native/libs/System.Native/pal_networkchange.c:
```diff
@@ -175,8 +178,6 @@ Error SystemNative_ReadEvents(intptr_t sock, NetworkChangeEvent onNetworkChange)
-    // unreachable
-    abort();
-    return Error_SUCCESS;
+    return Error_EINVAL;
```
Should we keep the old aborting code as a fallback and add separate #if for HAIKU instead? I am not familiar enough with the contract of this function though.

trungnt2910 replied:
I've added back the `abort()` logic, as an assertion that the function is unreachable.

This is true because on unsupported platforms, the higher-level C# libraries should already throw while trying to construct a network events socket.

Returning `Error_SUCCESS` does not make much sense though.

trungnt2910 commented on src/coreclr/minipal/Unix/doublemapping.cpp:
```diff
@@ -53,14 +53,24 @@ bool VMToOSInterface::CreateDoubleMemoryMapper(void** pHandle, size_t *pMaxExecu
-    name[sizeof(name) - 1] = '\0';
-    shm_unlink(name);
-    int fd = shm_open(name, O_RDWR | O_CREAT | O_EXCL | O_NOFOLLOW, 0600);
 #else // TARGET_FREEBSD
```
Done. The final `#else` assigns the `fd` variable to `-1` instead of doing the POSIX fallback itself, in case any of the platform-specific functions above return an error.

This should solve #72192 from the .NET 7 era.

trungnt2910 commented:
Strange, I'm not modifying any FreeBSD-related codepaths, what's wrong with the FreeBSD build?

jkotas commented:
FreeBSD build error is:
```
/__w/1/s/src/coreclr/pal/src/misc/sysinfo.cpp:215:57: error: use of undeclared identifier 'NUPML4E'
    lpSystemInfo->lpMaximumApplicationAddress = (PVOID) VM_MAXUSER_ADDRESS;
                                                        ^
/crossrootfs/x64/usr/include/machine/vmparam.h:186:35: note: expanded from macro 'VM_MAXUSER_ADDRESS'
#define VM_MAXUSER_ADDRESS      UVADDR(NUPML4E, 0, 0, 0)
                                       ^
```
I guess you have introduced by changing conditions around some of the includes.

trungnt2910 commented:
This is suspicious:

```
-- Looking for include file sys/user.h
-- Looking for include file sys/user.h - not found
```

I wonder how I could get more detailed CMake logs for the CI runs.

trungnt2910 commented:
When looking at other files, it seems like `sys/user.h` is only used on FreeBSD. Somehow `sysinfo.cpp` includes this header for all platforms.

This time, instead of checking for this header during configuration as in #86391, a FreeBSD `#ifdef` guard is added instead.

huoyaoyuan commented on src/coreclr/minipal/Unix/doublemapping.cpp:
```diff
@@ -53,15 +53,22 @@ bool VMToOSInterface::CreateDoubleMemoryMapper(void** pHandle, size_t *pMaxExecu
+    int fd = -1;
+#endif
+
+    // POSIX fallback
+    if (fd == -1)
+    {
+        char name[24];
+        sprintf(name, "/shm-dotnet-%d", getpid());
+        name[sizeof(name) - 1] = '\0';
+        shm_unlink(name);
+        fd = shm_open(name, O_RDWR | O_CREAT | O_EXCL | O_NOFOLLOW, 0600);
+        shm_unlink(name);
+    }
```
Shouldn't this block be in `#else`?

trungnt2910 replied:
No, it should not, as stated in https://github.com/dotnet/runtime/pull/93907#discussion_r1371114236.

Even for OSes with platform-specific ways to get a special memory file open, should their method fail, we can still always use the POSIX fallback.

huoyaoyuan commented on src/coreclr/gc/unix/numasupport.cpp:
```diff
@@ -9,9 +9,12 @@
-#include <sys/syscall.h>
 #include <minipal/utils.h>
 
+#ifdef TARGET_LINUX
+#include <sys/syscall.h>
+#endif
```
Is this used by other Unix systems?

trungnt2910 replied:
No, it is only used on Linux:

https://github.com/dotnet/runtime/blob/e5c631d28b985b773b28d0cb2346f49a92bd41dd/src/coreclr/gc/unix/numasupport.cpp#L52-L63

and 

https://github.com/dotnet/runtime/blob/e5c631d28b985b773b28d0cb2346f49a92bd41dd/src/coreclr/gc/unix/numasupport.cpp#L81-L83

huoyaoyuan commented on src/coreclr/pal/inc/pal_mstypes.h:
```diff
@@ -574,8 +574,8 @@ typedef long ptrdiff_t;
 typedef int ptrdiff_t;
 #endif // !HOST_64BIT
-#endif // !PAL_STDCPP_COMPAT
 #define _SIZE_T_DEFINED
+#endif // !PAL_STDCPP_COMPAT
```
What's the impact range of this?

trungnt2910 replied:
What do you mean by "impact range"?

The line should affect all platforms. However, I believe that this is how things should be, since with `PAL_STDCPP_COMPAT` the header should not define things that potentially mess up system headers.

Some more comments about this change here: https://github.com/dotnet/runtime/issues/55803#issuecomment-1537114390

trungnt2910 commented:
Anything known issues that I have to take extra attention of related to the Linux SSL implementation? Somehow I've been getting CI errors for this PR on Linux only; I don't really know what's happening.

mangod9 commented:
Since this is a large change, would it be possible to split the PR info smaller component PRs for PAL, hosting, libraries etc. so relevant teams can review the changes? Thx

trungnt2910 commented:
Not right now (I'm too busy). I'm still tracking it though, so if there are small specific issues I can resolve, I can fix it.

jkotas commented:
As @mangod9 suggested above, this needs to be split into multiple smaller PRs per area to make it reviewable. I am going to switch it to draft for now.

trungnt2910 committed:
Haiku: Add Haiku support for CoreCLR/PAL

This contains a part of the code required to build CoreCLR and get
`paltests` to pass on Haiku.

This commit covers `src/coreclr/pal/**`.

Co-authored-by: Jessica Hamilton <jessica.l.hamilton@gmail.com>

trungnt2910 commented:
I have edited this pull request to include only files in `src/coreclr/pal`.
Other parts will come in other PRs.
Please let me know if this needs to be split further.

trungnt2910 commented:
Latest builds are failing with this error:

```
  [  7%] Building CXX object inc/CMakeFiles/corguids.dir/__/pal/prebuilt/idl/cordebug_i.cpp.o
  In file included from /home/runner/work/dotnet-builds/dotnet-builds/runtime/src/coreclr/pal/inc/rt/palrt.h:630,
                   from /home/runner/work/dotnet-builds/dotnet-builds/runtime/src/coreclr/pal/inc/rt/rpc.h:15,
                   from /home/runner/work/dotnet-builds/dotnet-builds/runtime/src/coreclr/pal/prebuilt/idl/cordebug_i.cpp:29:
  /home/runner/work/dotnet-builds/dotnet-builds/runtime/src/coreclr/pal/inc/rt/safecrt.h:1093:16: error: conflicting declaration of C function 'size_t wcsnlen(const WCHAR*, size_t)'
   1093 | size_t __cdecl wcsnlen(const WCHAR *inString, size_t inMaxSize);
        |                ^~~~~~~
  In file included from /home/runner/work/dotnet-builds/dotnet-builds/dotnet-rootfs/boot/system/develop/headers/posix/wctype.h:10,
                   from /home/runner/work/dotnet-builds/dotnet-builds/runtime/src/coreclr/pal/inc/pal.h:51,
                   from /home/runner/work/dotnet-builds/dotnet-builds/runtime/src/coreclr/pal/inc/rt/palrt.h:136:
  /home/runner/work/dotnet-builds/dotnet-builds/dotnet-rootfs/boot/system/develop/headers/posix/wchar.h:122:17: note: previous declaration 'size_t wcsnlen(const wchar_t*, size_t)'
    122 | extern size_t   wcsnlen(const wchar_t *wcs, size_t maxLength);
        |                 ^~~~~~~
  /home/runner/work/dotnet-builds/dotnet-builds/runtime/src/coreclr/pal/inc/rt/safecrt.h:1098:16: error: conflicting declaration of C function 'size_t wcsnlen(const WCHAR*, size_t)'
   1098 | size_t __cdecl wcsnlen(const WCHAR *inString, size_t inMaxSize)
        |                ^~~~~~~
  /home/runner/work/dotnet-builds/dotnet-builds/dotnet-rootfs/boot/system/develop/headers/posix/wchar.h:122:17: note: previous declaration 'size_t wcsnlen(const wchar_t*, size_t)'
    122 | extern size_t   wcsnlen(const wchar_t *wcs, size_t maxLength);
        |                 ^~~~~~~
  make[2]: *** [inc/CMakeFiles/corguids.dir/build.make:76: inc/CMakeFiles/corguids.dir/__/pal/prebuilt/idl/cordebug_i.cpp.o] Error 1
  make[1]: *** [CMakeFiles/Makefile2:3250: inc/CMakeFiles/corguids.dir/all] Error 2
  make[1]: *** Waiting for unfinished jobs....
```

Seems like whoever included [`wctype.h`](https://github.com/dotnet/runtime/blob/9fa12355097e1ac17ac44445649a381323e9c35f/src/coreclr/pal/inc/pal.h#L51) did not expect `wchar.h` to be included as well.

According to the [POSIX Standard](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/wctype.h.html), however:

> [[CX](javascript:open_code('CX'))] [Option Start] Inclusion of the <wctype.h> header may make visible all symbols from the headers [<ctype.h>](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/ctype.h.html), [<stdarg.h>](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/stdarg.h.html), [<stddef.h>](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/stddef.h.html), [<stdio.h>](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/stdio.h.html), [<stdlib.h>](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/stdlib.h.html), [<string.h>](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/string.h.html), [<time.h>](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/time.h.html), and [<wchar.h>](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/wchar.h.html). [Option End]

What should I do in this case? Removing [`<wchar.h>`](https://github.com/haiku/haiku/blob/9f3bdf3d039430b5172c424def20ce5d9f7367d4/headers/posix/wctype.h#L10) from Haiku's system include just breaks everything. And according to POSIX Haiku isn't doing anything wrong in the first place...

trungnt2910 commented:
Hasn't broken my local Linux build and moved my local Haiku build to a different error. Let the CI inform the rest of the answer.

trungnt2910 commented:
The next Haiku error is:

```
  asmparse.y:176: error: "_IMPORT" redefined [-Werror]
  In file included from /home/trung/dotnet-rootfs/boot/system/develop/headers/posix/sys/types.h:11,
                   from /home/trung/dotnet-rootfs/boot/system/develop/headers/posix/stdio.h:9,
                   from /home/trung/dotnet-rootfs/boot/system/develop/headers/bsd/stdio.h:9,
                   from /home/trung/dotnet-runtime/src/coreclr/pal/inc/pal.h:39,
                   from /home/trung/dotnet-runtime/src/coreclr/pal/inc/rt/palrt.h:136,
                   from /home/trung/dotnet-runtime/src/coreclr/pal/inc/rt/rpc.h:15,
                   from /home/trung/dotnet-runtime/src/coreclr/pal/inc/rt/objidl.h:12,
                   from /home/trung/dotnet-runtime/src/coreclr/pal/inc/rt/ole2.h:8,
                   from /home/trung/dotnet-runtime/src/coreclr/inc/cor.h:16,
                   from /home/trung/dotnet-runtime/src/coreclr/ilasm/./ilasmpch.h:8,
                   from asmparse.y:9:
  /home/trung/dotnet-rootfs/boot/system/develop/headers/os/BeBuild.h:95: note: this is the location of the previous definition
     95 | #define _IMPORT
        |
  asmparse.y:262: error: "_EXPORT" redefined [-Werror]
  /home/trung/dotnet-rootfs/boot/system/develop/headers/os/BeBuild.h:90: note: this is the location of the previous definition
     90 | # define _EXPORT __attribute__((visibility("default")))
        |
```

Currently I am solving this by adding these lines to [`ilasmpch.h`](https://github.com/dotnet/runtime/blob/017593d90781d797df8b2241f6d1f83c236c442b/src/coreclr/ilasm/ilasmpch.h):

```diff
#include "mdfileformat.h"
#include "stgpooli.h"

+#ifdef _EXPORT
+#undef _EXPORT
+#endif

+#ifdef _IMPORT
+#undef _IMPORT
+#endif

#endif
```

I wonder whether these lines should be put directly to `pal.h` instead.

mangod9 commented:
Hey @trungnt2910, what is the status of this PR. Are you continuing to work on it?

trungnt2910 commented:
There's still an unanswered question from me above. Otherwise I believe the PR is ready for review.

mangod9 commented:
ok, so the question is about whether to add the undefs in `pal.h`? We are also triaging PRs for .NET 9, assume this can wait a few weeks ?

trungnt2910 commented:
Yes, the question is related to where I should add the `#undef`s, or otherwise how I should fix the redefined error.
I believe that one of the uses of `pal.h` is to isolate parts of CoreCLR code from system headers, so I initially thought putting those macros there would work.

> assume this can wait a few weeks ?

Yeah. This has been waiting for quite a few months now, waiting a few more weeks wouldn't hurt.

mangod9 commented:
adding @janvorli on advise on #undefs in pal.h.

janvorli commented:
> I wonder whether these lines should be put directly to `pal.h` instead.

This is strictly ilasm specific, so let's keep it where you have put it.

janvorli reviewed (commented):
Besides the two comments, it all looks good.

trungnt2910 committed:
ilasm: Haiku build fixes

Some OSes like Haiku define the `_EXPORT` and `_IMPORT` marcos.

Symbols beginning with an underscore followed immediately by an
uppercase letter are reserved anyway.

Add the guards to get them out of the way.

trungnt2910 committed:
pal: Remove usage of MAP_FILE

This flag is a compatibility flag and not required on most OSes.
Some, including Haiku, does not even define it.

trungnt2910 commented on src/coreclr/pal/src/include/pal/thread.hpp:
```diff
@@ -745,6 +744,7 @@ extern PAL_SafeActivationCheckFunction g_safeActivationCheckFunction;
 
 --*/
 #if defined(__linux__)
+#include <sys/syscall.h>
```
This is no longer applicable after the rebase.

trungnt2910 commented on src/coreclr/pal/src/map/map.cpp:
```diff
@@ -108,6 +108,11 @@ MAPmmapAndRecord(
 
+/* Some platforms like Haiku does not provide MAP_FILE */
+#ifndef MAP_FILE
+#define MAP_FILE 0
```
Done for `pal`. There are a few other usage instances across the `*AppBuilder`s, but that probably won't affect Haiku.

janvorli commented on src/native/minipal/debugger.c:
```diff
@@ -129,6 +131,14 @@ bool minipal_is_native_debugger_present(void)
+    {
+        return info.debugger_nub_thread > 0;
+    }
+    return FALSE;
```
I am sorry for missing it before, this should return "false" instead of "FALSE"

trungnt2910 replied:
Done.

I overlooked that, thought it was just a simple copy-paste...

janvorli reviewed (approved):
LGTM, thank you!

janvorli commented:
The outerloop failure is a known crossgen2 issue #104340
